From 919058f6ce87205863122563a30dfec2db4f9b20 Mon Sep 17 00:00:00 2001
From: Baha Mesleh <baha.mesleh@nokia.com>
Date: Fri, 10 Dec 2021 16:49:23 +0000
Subject: [PATCH 05/12] octeontx2-bphy-netdev: Fix cpri rx packet handling

commit 875916f58f5af5605cfe89ae5c9ffa264fff800a from
git@git.assembla.com:cavium/WindRiver.linux.git

The NAPI rx packet handler is not advancing sw_rd_ptr when
there is any error found. As a result, the same packet is
is processed in NAPI processing loop. This patch fixes the
issue.

Change-Id: I11a7e0d0da7b08df47360f838ad2330daca2bc25
Signed-off-by: Baha Mesleh <baha.mesleh@nokia.com>
Signed-off-by: Naveen Mamindlapalli <naveenm@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/69038
Reviewed-by: Sunil Kovvuri Goutham <sgoutham@marvell.com>
Tested-by: Sunil Kovvuri Goutham <sgoutham@marvell.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 .../ethernet/marvell/octeontx2/bphy/otx2_cpri.c  | 16 +++++++---------
 1 file changed, 7 insertions(+), 9 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_cpri.c b/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_cpri.c
index 536db19eb257..2fda900e22c9 100644
--- a/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_cpri.c
+++ b/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_cpri.c
@@ -165,8 +165,7 @@ static int otx2_cpri_process_rx_pkts(struct otx2_cpri_ndev_priv *priv,
 					    wqe->mhab_id, wqe->lane_id);
 			priv->stats.rx_dropped++;
 			priv->last_rx_dropped_jiffies = jiffies;
-			processed_pkts++;
-			continue;
+			goto update_processed_pkts;
 		}
 		priv2 = netdev_priv(netdev);
 		if (wqe->fcserr || wqe->rsp_ferr || wqe->rsp_nferr) {
@@ -176,8 +175,7 @@ static int otx2_cpri_process_rx_pkts(struct otx2_cpri_ndev_priv *priv,
 					    ul_cfg->sw_rd_ptr);
 			priv2->stats.rx_dropped++;
 			priv2->last_rx_dropped_jiffies = jiffies;
-			processed_pkts++;
-			continue;
+			goto update_processed_pkts;
 		}
 		if (unlikely(!netif_carrier_ok(netdev))) {
 			net_err_ratelimited("%s {cpri%d lmac%d} link down, drop pkt\n",
@@ -185,8 +183,7 @@ static int otx2_cpri_process_rx_pkts(struct otx2_cpri_ndev_priv *priv,
 					    priv2->lmac_id);
 			priv2->stats.rx_dropped++;
 			priv2->last_rx_dropped_jiffies = jiffies;
-			processed_pkts++;
-			continue;
+			goto update_processed_pkts;
 		}
 
 		len = wqe->pkt_length;
@@ -206,8 +203,7 @@ static int otx2_cpri_process_rx_pkts(struct otx2_cpri_ndev_priv *priv,
 					    netdev->name);
 			priv->stats.rx_dropped++;
 			priv->last_rx_dropped_jiffies = jiffies;
-			processed_pkts++;
-			continue;
+			goto update_processed_pkts;
 		}
 
 		memcpy(skb->data, pkt_buf, len);
@@ -216,12 +212,14 @@ static int otx2_cpri_process_rx_pkts(struct otx2_cpri_ndev_priv *priv,
 
 		netif_receive_skb(skb);
 
+		priv2->last_rx_jiffies = jiffies;
+
+update_processed_pkts:
 		processed_pkts++;
 		ul_cfg->sw_rd_ptr++;
 		if (ul_cfg->sw_rd_ptr == ul_cfg->num_entries)
 			ul_cfg->sw_rd_ptr = 0;
 
-		priv2->last_rx_jiffies = jiffies;
 	}
 
 	if (processed_pkts)
-- 
2.34.1

